File: LeetCode_Algorithm/62(Unique Paths)/62.js
var uniquePaths = function(m, n) {
  // sol1
  // var arr = []
  // for (let i = 0 ; i < m; i++) {
  //   arr[i] = new Array(n)
  // }

  // sol2 81ms 63%
  var arr = Array.from(Array(m), ()=>Array(n).fill(1))
  for (let r=1; r < arr.length; r++){
    for (let c=1; c < arr[0].length; c++) {
      arr[r][c] = arr[r-1][c] + arr[r][c-1]
    }
  }
  return arr[m-1][n-1]
};

console.log(uniquePaths(3,7))

File: LeetCode_Algorithm/62(Unique Paths)/62.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # sol1 O(mn) 39ms 73%
        board = [[1] * n for _ in range(m)]
        for r in range(1, m):
            for c in range(1, n):
                board[r][c] = board[r-1][c] + board[r][c-1]
        return board[m-1][n-1]

        # sol2

        pass


print(Solution().uniquePaths(3,7))